### ISSUE TITLE
New keywords for compatibility: removed, deprecated

### ISSUE THREAD

**Issue by fklebert** [ndsev/zserio#486]
For the compatible use of schemas over versions, we would like to introduce some new keywords, usable for different zserio language elements:

We would like to have not only a deprecated keyword for enum values (that might even differ from the ones for structs in case you find it useful), but also a “removed” keyword for enum values.
 
Let’s make an example to better illustrate things:
 
```
enum uint8 Type {
    NONE,
    HEAVY,
    LIGHT,
    MID,
};
```
 
We will never use the dedicated value assignment like NONE = 1, unless it really has a special meaning.
The idea to not use the value assignment is that it enforces the schema to keep enum values, even if you later decide to delete e.g. HEAVY = 2.
If you delete it without the direct assignment, all of the values will shift and you can detect something like this very easily with a CI test.
There was of course discussion that if we use the dedicated assignment, that we can simply generate holes, which may come handy when we want to tidy up the schema. That would be backward compatible, but not forward compatible as new code would throw an error if it finds a removed value in the old data.
Also it bares the risk that you later reassign the underlying value 2 to a new enum symbol name, e.g. EXTRA_HEAVY = 2.
So old binaries will be interpreted wrongly.
 
So basically we are thinking about two different keywords for enum values:
 
_deprecated_: will **throw a warning** when new code writes it, reading will not fail
_removed_: will **throw an error** when new code writes it, reading will not fail
 
Maybe the schema could look like this:
 
```
enum uint8 Type {
    NONE = 1,
    HEAVY = 2 [[removed]],
    LIGHT = 3 [[deprecated]],
    MID = 4,
};
```
 
Of course any other notation will also work 😊

**Comment by Mi-La** [comment]
Our proposal for the new syntax (which is already more inspired by Java than C++) is to use decorator-like syntax for deprecated and removed.
```
enum uint8 Type {
    NONE = 1,
    @removed HEAVY = 2,
    @deprecated LIGHT = 3,
    MID = 4,
};
```
This could be extended to support more language elements in future - e.g. structure fields, bitmask fields, whole compounds etc.
> Note that @removed for structure fields would be much more complicated since fields can be used in various expressions (conditionals, constraints, parameters, ...) which could lead to many consequences.

Then we have idea how to define the behavior:
1. `deprecated` - we have no way to fire a warning on runtime. We just plan to use deprecated feature which should be somehow supported in all languages:
   * C++
      * `__attribute__(deprecated)` in gcc (enumerators supported from [gcc-6](https://gcc.gnu.org/gcc-6/changes.html)), MinGW and clang
      * `[[deprectated]]` supported from [MSVC 2015](https://learn.microsoft.com/en-us/cpp/cpp/attributes?view=msvc-140#deprecated)
      * note that since we still have only C++11, we cannot just use `[[deprecated]]` which is available for enumerators from [C++17](https://en.cppreference.com/w/cpp/language/attributes/deprecated)
   * Java [Deprecated](https://docs.oracle.com/javase/8/docs/api/java/lang/Deprecated.html)
   * Python has [DeprecationWarning](https://docs.python.org/3.8/library/exceptions.html#DeprecationWarning), maybe we can use [Deprecated](https://pypi.org/project/Deprecated/) package
      * note that this will cause a print-out on the console on runtime!
2. `removed` - we can throw an exception when a removed enum value is going to be written to the stream
   * moreover we can rename the removed enum item in generated code to something like `ZSERIO_REMOVED_HEAVY` which will prevent the user code of using the removed item by an accident, while it would be still possible to read
